### PR TITLE
tgui: Updates typescript, other dependencies

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
The largest change to this is that it updates typescript from 4.3.5 to 4.9.4.

It also does some QoL to update @ types, yarn version, and typescript-eslint-parser (OPTIONAL)

I did this by:
1. Updating typescript `yarn add typescript`
2. Updating other types `yarn add @types/...`
3. Repackaging typescript (and other) workplace sdks `yarn dlx @yarnpkg/sdks`
4. Updates yarn via corepack `yarn set version stable`
5. Fixes a couple places where typescript was mad

## Why It's Good For The Game
- I want to do things like make discriminating type unions
- Huge leap into typescript so I can better type out tgui core
- TS >>> JS

## Changelog
N/A Nothing player facing

### CODE REVIEW

File: tgui/.eslintrc.yml [hunk 1]
```diff
@@ -308,13 +308,16 @@ rules:
...
+  comma-dangle: [
+      error,
+      {
+        arrays: always-multiline,
+        objects: always-multiline,
+        imports: always-multiline,
+        exports: always-multiline,
+        functions: only-multiline, ## Optional on functions
+      },
+    ]
```
Comment (jlsnow301) [inline tgui/.eslintrc.yml:320]: I was getting errors for a badly formatted YAML entry, so this fixed it. I can revert this if it's not to your liking

**Comment by jlsnow301** [comment]
Could use a thorough review on this one @stylemistake

File: tgui/.vscode/extensions.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (stylemistake) [inline tgui/.vscode/extensions.json:1]: .vscode folder has no effect inside tgui folder, so it should be deleted, and any new settings should be added to the root .vscode folder.

File: tgui/package.json [hunk 3]
```diff
@@ -26,12 +26,12 @@
     "@babel/plugin-transform-jscript": "^7.14.5",
     "@babel/preset-env": "^7.15.0",
     "@babel/preset-typescript": "^7.15.0",
-    "@types/jest": "^27.0.1",
-    "@types/jsdom": "^16.2.13",
-    "@types/node": "^14.17.9",
+    "@types/jest": "^29.2.4",
+    "@types/jsdom": "^20.0.1",
+    "@types/node": "^18.11.18",
```
Comment (stylemistake) [inline tgui/package.json:31]: This should be pinned to v14

File: tgui/.eslintrc.yml (hunk 1 above)
Comment (stylemistake) [reply tgui/.eslintrc.yml:320]: Would be good if YAMLs were ignored by prettier.

File: tgui/.vscode/extensions.json (hunk 2 above)
Comment (jlsnow301) [reply tgui/.vscode/extensions.json:1]: Agree, but yarn kept insisting this exists when I run dlx for some reason

Comment (stylemistake) [reply tgui/.vscode/extensions.json:1]: It doesn't know where vscode root is, and this one will be unused, so it's safe to delete.

File: tgui/packages/tgui/assets.ts [hunk 4]
```diff
@@ -9,7 +9,17 @@ const loadedMappings = {};
 
 export const resolveAsset = (name) => loadedMappings[name] || name;
 
-export const assetMiddleware = (store) => (next) => (action) => {
+type Action =
```
Comment (jlsnow301) [inline tgui/packages/tgui/assets.ts:12]: here's your discriminating type union, working!

File: tgui/packages/tgui/assets.ts [hunk 5]
```diff
@@ -9,7 +9,17 @@ const loadedMappings = {};
...
+  | {
+      type: 'asset/stylesheet';
+      payload: string;
+    }
+  | {
+      type: 'asset/mappings';
+      payload: Record<string, string>;
+    };
```
Comment (jlsnow301) [inline tgui/packages/tgui/assets.ts:20]: here's the discriminating type union! neat

**Review by stylemistake — APPROVED** [review]

**Pull request merged**